Simplenote/src/main/java/com/automattic/simplenote/utils/SyncWorker.java

```diff
@@ -0,0 +1,107 @@
+package com.automattic.simplenote.utils;
+
+import android.content.Context;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.work.Worker;
+import androidx.work.WorkerParameters;
+
+import com.automattic.simplenote.Simplenote;
+import com.automattic.simplenote.models.Note;
+import com.automattic.simplenote.models.Preferences;
+import com.automattic.simplenote.models.Tag;
+import com.automattic.simplenote.utils.AppLog.Type;
+import com.simperium.client.Bucket;
+
+import static com.automattic.simplenote.Simplenote.TEN_SECONDS_MILLIS;
+
+public class SyncWorker extends Worker {
+    private Bucket<Note> mBucketNote;
+    private Bucket<Preferences> mBucketPreference;
+    private Bucket<Tag> mBucketTag;
+
+    public SyncWorker(@NonNull Context context, @NonNull WorkerParameters params) {
+        super(context, params);
+        Simplenote application = (Simplenote) context.getApplicationContext();
+        mBucketNote = application.getNotesBucket();
+        mBucketTag = application.getTagsBucket();
+        mBucketPreference  = application.getPreferencesBucket();
+    }
+
+    @NonNull
+    @Override
+    public Result doWork() {
+        AppLog.add(Type.NETWORK, NetworkUtils.getNetworkInfo(getApplicationContext()));
+
+        if (mBucketNote != null) {
+            mBucketNote.start();
+            AppLog.add(Type.SYNC, "Started note bucket (SyncWorker)");
+        }
+
+        if (mBucketTag != null) {
+            mBucketTag.start();
+            AppLog.add(Type.SYNC, "Started tag bucket (SyncWorker)");
+        }
+
+        if (mBucketPreference != null) {
+            mBucketPreference.start();
+            AppLog.add(Type.SYNC, "Started preference bucket (SyncWorker)");
+        }
+
+        Log.d("SyncWorker.doWork", "Started buckets");
+
+        new Handler(Looper.getMainLooper()).postDelayed(
+            new Runnable() {
+                @Override
+                public void run() {
+                    if (((Simplenote) getApplicationContext()).isInBackground()) {
+                        if (mBucketNote != null) {
+                            mBucketNote.stop();
+                            AppLog.add(Type.SYNC, "Stopped note bucket (SyncWorker)");
+                        }
+
+                        if (mBucketTag != null) {
+                            mBucketTag.stop();
+                            AppLog.add(Type.SYNC, "Stopped tag bucket (SyncWorker)");
+                        }
+
+                        if (mBucketPreference != null) {
+                            mBucketPreference.stop();
+                            AppLog.add(Type.SYNC, "Stopped preference bucket (SyncWorker)");
+                        }
+
+                        Log.d("SyncWorker.doWork", "Stopped buckets");
+                    }
+                }
+            },
+            TEN_SECONDS_MILLIS
+        );
+
+        return Result.retry();
```

Comment:
AFAIU this might not work as intended - the OS keeps the device awake and gives the app a higher priority only while the "doWork" method is being executed. If we return from the method right away we'll lose these privileges. Moreover, the system can kill/destroy the worker anytime. 

From the docs:
"Worker classes are instantiated at runtime by WorkManager and the doWork() method is called on a pre-specified background thread (see Configuration.getExecutor()). This method is for synchronous processing of your work, meaning that once you return from that method, the Worker is considered to be finished and will be destroyed."

Wdyt?

Reply:
Which part do you think will not work as intended?  Are you talking about the post-delayed handler stopping the buckets?  When testing these changes, the buckets were stopped as expected.

```
2020-08-25 11:07:01.796 17642-17642/com.automattic.simplenote.debug D/Simplenote.onActivityPaused: Started worker
2020-08-25 11:07:11.866 17642-17883/com.automattic.simplenote.debug D/SyncWorker.doWork: Started buckets
2020-08-25 11:07:21.908 17642-17642/com.automattic.simplenote.debug D/SyncWorker.doWork: Stopped buckets
2020-08-25 11:08:11.912 17642-17991/com.automattic.simplenote.debug D/SyncWorker.doWork: Started buckets
2020-08-25 11:08:21.954 17642-17642/com.automattic.simplenote.debug D/SyncWorker.doWork: Stopped buckets
2020-08-25 11:08:24.860 17642-17642/com.automattic.simplenote.debug D/Simplenote.onActivityResumed: Stopped worker
```

Reply:
I think both the syncing itself and stop() might not work as intended. When the worker finishes the OS assumes the app finished the job.

The fact that it works during our tests unfortunately doesn't prove anything. Perhaps the issue occurs only when the worker is started from a doze mode or on low batter or something else. It's hard to tell. However, the documentation is quite specific about this. Wdyt?

Reply:
>What issue?

After the worker is done - the `doWork` method completes - the system can kill the app -> it can kill even the syncing. Tbh I don't see any reason why we should go directly against the documentation. I'm also worried that other devs might copy this implementation (about which we already know that sometime doesn't work) and use it for other jobs which can't be interrupted in the middle.

Reply:
We're not intentionally going against the documentation.

> about which we already know that sometime doesn't work

What do you mean?

Reply:
How come? The documentation says `This method is for synchronous processing of your work`, but we are starting an asynchronous work there.

Reply:
Ah, I see what you mean.  The post-delayed handler and the buckets by their nature are asynchronous.